### PR TITLE
Use `Identity::from_pem`

### LINKED ISSUE #1731: Agent should read private.pem and self.cert instead of creating identy.pfx
Instead of calling out to create a tertiary file, private.pem and self.cert should be read into a single buffer and passed to `Identity::from_pem()` instead of using `from_pkcs12_der()`

https://github.com/whamcloud/integrated-manager-for-lustre/blob/675374cd47f709d1a3d59381fa6459d4329362f2/iml-agent/src/http_comms/crypto_client.rs#L16-L18

Related to Pull #1730

### PR DESCRIPTION
**Pull request by mkpankov**
Close #1731

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2054)

### CODE REVIEW

File: iml-agent/src/env.rs [hunk 1]
```diff
@@ -41,11 +32,11 @@ fn get_cert_path() -> String {
     get_var("CRT_PATH")
 }
 
-fn get_pfx_path() -> String {
+fn _get_pfx_path() -> String {
```
Comment (ip1981) [inline iml-agent/src/env.rs:35]: If it's unused, why not delete it?

File: iml-agent/src/env.rs [hunk 2]
```diff
@@ -59,43 +50,24 @@ pub fn mailbox_sock(mailbox: &str) -> String {
...
+    pub static ref PEM: Vec<u8> = {
+        let mut result = Vec::new();
+
         let private_pem_path = get_private_pem_path();
 
-        if !path_exists(&private_pem_path) {
-            panic!("{} does not exist", private_pem_path)
-        };
+        let mut private_pem = File::open(private_pem_path)
+            .unwrap_or_else(|_| panic!("{} does not exist", get_private_pem_path()));
```
Comment (ip1981) [inline iml-agent/src/env.rs:59]: Technically, I think, there could be other errors than "not exist".
It would probably be better to sue the argument of `unwrap_or_else` and print actual error.

Also maybe use `private_pem_path` instead of calling `get_private_pem_path()`?

File: iml-agent/src/env.rs [hunk 3]
```diff
@@ -59,43 +50,24 @@ pub fn mailbox_sock(mailbox: &str) -> String {
...
-                &authority_cert_path,
-                "-passout",
-                "pass:",
-            ])
-            .status()
-            .expect("Error creating pfx");
-
-        std::fs::read(&pfx_path).expect("Could not read pfx")
+        let mut cert =
+            File::open(cert_path).unwrap_or_else(|_| panic!("{} does not exist", get_cert_path()));
```
Comment (ip1981) [inline iml-agent/src/env.rs:67]: Likewise here.

**Comment by mkpankov** [comment]
@jgrund making sure this works with a "proper" certificate proves to be harder than I thought.

I can't use LetsEncrypt because they want a public-facing web-server or DNS record, which doesn't make sense for our vagrant cluster or the isolated production environment it mimics.

I currently think the only way to actually setup this in production is to use self-signed certificates.

At one of my previous employers we had corporate certificates derived from custom Root CA, and that root CA had to be installed manually on each machine and browser separately to add it to chain of trust. So that root CA was self-signed, and I don't think using non-self-signed certificate for this kind of setup makes sense.

Point is, I'm not sure anyone buys proper certificates from 3-rd party public authority for TLS on *intranet*. In that's actually what we want to support, I guess we should get a proper certificate and PK from a provider like digicert.

**Comment by ip1981** [comment]
There is no suck thing like a "proper" certificate (except algorithms). A certificate is either trusted or not (via a chain of certificates, also known as intermediate certificates). That is whether a certificate is signed by a certificate which is signed by a certificate and so on ... by a certificate which you trust :)

**Comment by mkpankov** [comment]
I'm saying the same, so it's not clear what we wanted to test with "not a self-signed client certificate". We aren't expecting our clients to obtain certificates with publicly trusted root CA (which are usually paid, not free), are we?

**Comment by mkpankov** [comment]
I've removed `danger_allow_invalid_certificates` from `ClientBuilder` invocation and the tests pass, so this should work in all scenarios.

File: iml-agent/src/http_comms/crypto_client.rs [hunk 4]
```diff
@@ -25,7 +25,7 @@ pub fn get_id(pfx: &[u8]) -> Result<Identity, ImlAgentError> {
 /// * `id` - The client identity to use
 pub fn create_client(id: Identity) -> Result<Client, ImlAgentError> {
     Client::builder()
-        .danger_accept_invalid_certs(true)
```
Comment (jgrund) [inline iml-agent/src/http_comms/crypto_client.rs:28]: Can you split this change out to it's own patch?

**Comment by johnsonw** [comment]
Hi @mkpankov,

The integration tests are picking up the following error from the rust-iml-agent on the storage servers when detecting the filesystem :

```
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.472#033[0m #033[33m WARN#033[0m rustls::session: Sending fatal alert BadCertificate
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.472#033[0m #033[33m WARN#033[0m iml_agent::reader: Got a manager read Error reqwest::Error { kind: Request, url: "https://adm.local/agent2/message/?server_boot_time=2020-07-13T10%3A09%3A13.000000%2B00%3A00Z&client_start_time=2020-07-13T10%3A09%3A53.579173%2B00%3A00Z", source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) } }) }. Will retry in 5 seconds.
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.472#033[0m #033[32m INFO#033[0m iml_agent::http_comms::session: Terminating all sessions
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.603#033[0m #033[32m INFO#033[0m iml_agent::poller: sending session create request for ntp
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::http_comms::agent_client: Requesting new session for: PluginName("ntp").
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::poller: sending session create request for action_runner
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::http_comms::agent_client: Requesting new session for: PluginName("action_runner").
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::poller: sending session create request for ostpool
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::http_comms::agent_client: Requesting new session for: PluginName("ostpool").
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::poller: sending session create request for stats
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::http_comms::agent_client: Requesting new session for: PluginName("stats").
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::poller: sending session create request for device
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::http_comms::agent_client: Requesting new session for: PluginName("device").
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::poller: sending session create request for postoffice
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.604#033[0m #033[32m INFO#033[0m iml_agent::http_comms::agent_client: Requesting new session for: PluginName("postoffice").
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.616#033[0m #033[33m WARN#033[0m rustls::session: Sending fatal alert BadCertificate
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.616#033[0m #033[33m WARN#033[0m rustls::session: Sending fatal alert BadCertificate
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.616#033[0m #033[32m INFO#033[0m iml_agent::poller: session create request for postoffice failed: Reqwest(reqwest::Error { kind: Request, url: "https://adm.local/agent2/message/", source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) } }) })
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.616#033[0m #033[32m INFO#033[0m iml_agent::poller: session create request for device failed: Reqwest(reqwest::Error { kind: Request, url: "https://adm.local/agent2/message/", source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) } }) })
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.618#033[0m #033[33m WARN#033[0m rustls::session: Sending fatal alert BadCertificate
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.618#033[0m #033[32m INFO#033[0m iml_agent::poller: session create request for ostpool failed: Reqwest(reqwest::Error { kind: Request, url: "https://adm.local/agent2/message/", source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) } }) })
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.618#033[0m #033[33m WARN#033[0m rustls::session: Sending fatal alert BadCertificate
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.619#033[0m #033[32m INFO#033[0m iml_agent::poller: session create request for ntp failed: Reqwest(reqwest::Error { kind: Request, url: "https://adm.local/agent2/message/", source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) } }) })
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.619#033[0m #033[33m WARN#033[0m rustls::session: Sending fatal alert BadCertificate
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.619#033[0m #033[32m INFO#033[0m iml_agent::poller: session create request for stats failed: Reqwest(reqwest::Error { kind: Request, url: "https://adm.local/agent2/message/", source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) } }) })
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.625#033[0m #033[33m WARN#033[0m rustls::session: Sending fatal alert BadCertificate
Jul 13 10:15:39 oss1 iml-agent-daemon: #033[2mJul 13 10:15:39.625#033[0m #033[32m INFO#033[0m iml_agent::poller: session create request for action_runner failed: Reqwest(reqwest::Error { kind: Request, url: "https://adm.local/agent2/message/", source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: InvalidData, error: WebPKIError(UnknownIssuer) } }) })
```

**Comment by mkpankov** [comment]
@johnsonw thanks Will. Did you go to the machine to see that or it's possible to see in Jenkins logs?

File: iml-agent/src/http_comms/crypto_client.rs (hunk 4 above)
Comment (mkpankov) [reply iml-agent/src/http_comms/crypto_client.rs:28]: Done
https://github.com/whamcloud/integrated-manager-for-lustre/pull/2067

**Comment by mkpankov** [comment]
I've split the patches and this branch passes the tests.

**Comment by jgrund** [comment]
There are a few outstanding comments, otherwise LGTM

**Comment by johnsonw** [comment]
> @johnsonw thanks Will. Did you go to the machine to see that or it's possible to see in Jenkins logs?

I was able to get the information from the sos reports (part of the build artifacts)

File: iml-agent/src/env.rs (hunk 1 above)
Comment (mkpankov) [inline iml-agent/src/env.rs]: Done

**Comment by mkpankov** [comment]
Follow-up issue: #2071
